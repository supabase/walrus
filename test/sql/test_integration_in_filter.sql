select 1 from pg_create_logical_replication_slot('realtime', 'wal2json', false);

create table public.notes(
    id int primary key,
    body text
);

alter table public.notes replica identity full;

insert into realtime.subscription(subscription_id, entity, claims, filters)
select
    seed_uuid(id),
    'public.notes',
    jsonb_build_object(
        'role', 'authenticated',
        'email', 'example@example.com',
        'sub', seed_uuid(id)::text
    ),
    array[(column_name, op, value)::realtime.user_defined_filter]
from
    (
        values
            (1 , 'body', 'in', array['aaa', 'bbb', 'ccc']::text),
            (2 , 'body', 'in', array['aaa', 'ccc']::text),
            (3 , 'body', 'in', array[]::text[]::text)

    ) f(id, column_name, op, value);


select clear_wal();
insert into public.notes(id, body) values (1, 'bbb');

delete from public.notes;

select subscription_id, filters from realtime.subscription;

select
    rec,
    is_rls_enabled,
    subscription_ids,
    errors
from
   walrus;

-- Confirm that filtering on `in` more than 100 entries throws an error
insert into realtime.subscription(subscription_id, entity, claims, filters)
select
    seed_uuid(6),
    'public.notes',
    jsonb_build_object(
        'role', 'authenticated',
        'email', 'example@example.com',
        'sub', seed_uuid(6)::text
    ),
    array[('body', 'in', array[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1])::realtime.user_defined_filter];

drop table public.notes;
select pg_drop_replication_slot('realtime');
truncate table realtime.subscription;
