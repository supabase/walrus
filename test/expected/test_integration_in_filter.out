select 1 from pg_create_logical_replication_slot('realtime', 'wal2json', false);
 ?column? 
----------
        1
(1 row)

create table public.notes(
    id int primary key,
    body text
);
insert into realtime.subscription(subscription_id, entity, claims, filters)
select
    seed_uuid(id),
    'public.notes',
    jsonb_build_object(
        'role', 'authenticated',
        'email', 'example@example.com',
        'sub', seed_uuid(id)::text
    ),
    array[(column_name, op, value)::realtime.user_defined_filter]
from
    (
        values
            (1 , 'body', 'in', array['aaa', 'bbb', 'ccc']::text),
            (2 , 'body', 'in', array['aaa', 'ccc']::text),
            (3 , 'body', 'in', array[]::text[]::text)
    ) f(id, column_name, op, value);
select clear_wal();
 clear_wal 
-----------
 
(1 row)

insert into public.notes(id, body) values (1, 'bbb');
delete from public.notes;
select subscription_id, filters from realtime.subscription;
           subscription_id            |             filters             
--------------------------------------+---------------------------------
 f4539ebe-c779-5788-bbc1-2421ffaa8954 | {"(body,in,\"{aaa,bbb,ccc}\")"}
 5211e8ec-8c25-5c7f-9b03-6ff1eac0159e | {"(body,in,\"{aaa,ccc}\")"}
 11955172-4e1d-5836-925f-2bcb7a287b87 | {"(body,in,{})"}
(3 rows)

select
    rec,
    is_rls_enabled,
    subscription_ids,
    errors
from
   walrus;
                        rec                         | is_rls_enabled |            subscription_ids            | errors 
----------------------------------------------------+----------------+----------------------------------------+--------
 {                                                 +| f              | {f4539ebe-c779-5788-bbc1-2421ffaa8954} | {}
     "type": "INSERT",                             +|                |                                        | 
     "table": "notes",                             +|                |                                        | 
     "record": {                                   +|                |                                        | 
         "id": 1,                                  +|                |                                        | 
         "body": "bbb"                             +|                |                                        | 
     },                                            +|                |                                        | 
     "schema": "public",                           +|                |                                        | 
     "columns": [                                  +|                |                                        | 
         {                                         +|                |                                        | 
             "name": "id",                         +|                |                                        | 
             "type": "int4"                        +|                |                                        | 
         },                                        +|                |                                        | 
         {                                         +|                |                                        | 
             "name": "body",                       +|                |                                        | 
             "type": "text"                        +|                |                                        | 
         }                                         +|                |                                        | 
     ],                                            +|                |                                        | 
     "commit_timestamp": "2000-01-01T08:01:01.000Z"+|                |                                        | 
 }                                                  |                |                                        | 
 {                                                 +| f              | {}                                     | {}
     "type": "DELETE",                             +|                |                                        | 
     "table": "notes",                             +|                |                                        | 
     "schema": "public",                           +|                |                                        | 
     "columns": [                                  +|                |                                        | 
         {                                         +|                |                                        | 
             "name": "id",                         +|                |                                        | 
             "type": "int4"                        +|                |                                        | 
         },                                        +|                |                                        | 
         {                                         +|                |                                        | 
             "name": "body",                       +|                |                                        | 
             "type": "text"                        +|                |                                        | 
         }                                         +|                |                                        | 
     ],                                            +|                |                                        | 
     "old_record": {                               +|                |                                        | 
         "id": 1                                   +|                |                                        | 
     },                                            +|                |                                        | 
     "commit_timestamp": "2000-01-01T08:01:01.000Z"+|                |                                        | 
 }                                                  |                |                                        | 
(2 rows)

-- Confirm that filtering on `in` more than 100 entries throws an error
insert into realtime.subscription(subscription_id, entity, claims, filters)
select
    seed_uuid(6),
    'public.notes',
    jsonb_build_object(
        'role', 'authenticated',
        'email', 'example@example.com',
        'sub', seed_uuid(6)::text
    ),
    array[('body', 'in', array[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1])::realtime.user_defined_filter];
ERROR:  too many values for `in` filter. Maximum 100
CONTEXT:  PL/pgSQL function realtime.subscription_check_filters() line 48 at RAISE
drop table public.notes;
select pg_drop_replication_slot('realtime');
 pg_drop_replication_slot 
--------------------------
 
(1 row)

truncate table realtime.subscription;
